cmd/zc_traverser_blob.go

```diff
@@ -42,6 +42,39 @@ type blobTraverser struct {
 	incrementEnumerationCounter func()
 }
 
+func (t *blobTraverser) isDirectory() bool {
+	isDirDirect := copyHandlerUtil{}.urlIsContainerOrVirtualDirectory(t.rawURL)
+	if isDirDirect {
+		return isDirDirect
+	}
+
+	// To find out if it's a directory or a individual blob, let's try to list from it.
+	blobURLParts := blobURLPartsExtension{azblob.NewBlobURLParts(*t.rawURL)}
+
+	rawContainerURL := blobURLParts.getContainerURL()
+	searchPrefix := blobURLParts.BlobName
+
+	// If this is a single blob, ensure it doesn't get included in the search.
+	if !strings.HasSuffix(searchPrefix, "/") {
+		searchPrefix += "/"
+	}
+
+	containerURL := azblob.NewContainerURL(rawContainerURL, t.p)
+	resp, err := containerURL.ListBlobsFlatSegment(t.ctx, azblob.Marker{}, azblob.ListBlobsSegmentOptions{})
+
+	if err != nil {
+		return false
+	}
+
+	// If items exist in the search, this IS a virtual directory.
+	// Virtual directories aren't otherwise possible in blob storage.
+	if len(resp.Segment.BlobItems) > 0 {
```

Comment:
to clarify, we are verifying whether the path represents an **existing** virtual directory, right? if yes, we should update the function name accordingly

Reply:
Not anymore. This function is intended to check whether something is intended to be a directory on the destination, and whether something is an existing directory on the source.

Reply:
Any blob path can be a virtual directory though, they are called virtual because they don't really exist. So for blob, it should always return true then?

Reply:
Well, no. If a user wants to upload `filexyz` under `container/filexyz` and `container/filexyz/filexyz` (for whatever reason), in the past of azcopy, the command

`azcopy copy filexyz <URL>/container/filexyz` would treat the result as an immediate destination

and

`azcopy copy filexyz <URL>/container/filexyz/` would treat it as a virtual directory.

For uploads, this information is important because it determines where we place things.
For downloads, this information is important because it determines if we need to be recursive or not.

Reply:
I've yet to push my changes to github yet, still working on them locally, but you'll see what I mean after they come along.

Reply:
Ok, I see what you mean now, it matters when uploading a single file, because we need to know whether the blob should be put "under" the given dst path, or "at" the given path.